### ISSUE TITLE
Top-K outputs

### ISSUE THREAD

**Issue by lefnire** [Conchylicultor/DeepQA#17]
Thanks for creating this repo; very thorough & well-documented compared to other chat-bot samples. Apologies if I'm misusing the issue-queue; I've tried my hand at Google / arxiv.org over the past few days, finally breaking down to bother the source :)

I'd like to use this code similar to Allo, where you say a thing and the bot gives the top 3 candidate responses. You can either tap one or type your own response. "What's my bank balance?" -> "For which account?" -> {"Checking Account" | "Savings Account" | "I don't have one" | custom input} ([example](https://upload.wikimedia.org/wikipedia/en/8/8f/Google_Allo.jpg)). I _believe_ this is similar to what you're discussing in your top item under [Improvements](https://github.com/Conchylicultor/DeepQA#improvements), but the tech is a tad lost on me. I know enough TensorFlow take a stab; but less than needed to grok that bullet-point - so I'm wondering if you wouldn't mind a paragraph or two outlining how I might go about this? My shot-in-the-dark is using a different `seq2seq` model [here](https://github.com/Conchylicultor/DeepQA/blob/master/chatbot/model.py#L159) and a `loop_function` argument?

**Comment by Conchylicultor** [comment]
When the network predict a word,  what it does is to predict the softmax probability distribution for each words. For instance if I say `Hi`, the bot will compute the most likely words:

```
- Hi (34%)
- Hello (12%)
- How (7%)
- My (5%)
- ...
```

For now, to predict the next word, the program just take the previous word which has the highest probability score (here `hi` with 34%). The ideas would be instead to roll a dice and select the next word accoring to the softmax distribution. That mean you'll have 34% chance of choosing `hi`, 12% chance of choosing `hello`,...

In order to do that, some changes need to be done on the code. There the loop function parameters on the TensorFlow Rnn decoder which would allow to do that: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/seq2seq.py#L117
As I said previously, for now, the default embedding seq2seq model use argmax to simply take the most likely prediction:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/seq2seq.py#L274
That would have to be replaced by a custom function which sample the sofmax distribution. But in order to do that, `embedding_rnn_seq2seq` and `embedding_rnn_decoder` would have to be redefined. 

I hope this is clearer, let me know if it's not.

Their may gonna be a problem though, that the loop function would need a way to keep track of chosen samples otherwise it won't be possible to reconstruct the output. Indeed, the output of the decoder will always be the SoftMax distribution, here https://github.com/Conchylicultor/DeepQA/blob/master/chatbot/textdata.py#L441 the function need a way to know which outputs have been chosen by the loop functions.
That should not be to difficult to do though.

**Comment by lefnire** [comment]
Fantastic, this gives me some to work with. If I make progress, I'll fork & link here with the commits. Thanks a million!
